packages/marshal/marshal.js

```diff
@@ -564,3 +644,84 @@ export function makeMarshal(
     unserialize,
   });
 }
+
+/**
+ * Create and register a Remotable.  After this, getInterfaceOf(remotable)
+ * returns iface.
+ *
+ * // https://github.com/Agoric/agoric-sdk/issues/804
+ *
+ * @param {string} [iface='Remotable'] The interface specification for the remotable
+ * @param {object} [props={}] Own-properties are copied to the remotable
+ * @param {object} [remotable={}] The object used as the remotable
+ * @returns {object} remotable, modified for debuggability
+ */
+function Remotable(iface = 'Remotable', props = {}, remotable = {}) {
+  iface = pureCopy(iface);
+  const ifaceType = typeof iface;
+
+  // Find the alleged name.
+  if (ifaceType !== 'string') {
+    throw Error(`Interface must be a string, not ${ifaceType}; unimplemented`);
+  }
+
+  // Fail fast: check that the unmodified object is able to become a Remotable.
+  assertCanBeRemotable(remotable);
+
+  // Ensure that the remotable isn't already registered.
+  if (remotableToInterface.has(remotable)) {
+    throw Error(`Remotable ${remotable} is already mapped to an interface`);
+  }
+
+  // TODO: Get the allegedName a different way.
+  const allegedName = iface;
+
+  // A prototype for debuggability.
+  const oldRemotableProto = harden(Object.getPrototypeOf(remotable));
+
+  // Fail fast: create a fresh empty object with the old
+  // prototype in order to check it against our rules.
+  mustPassByPresence(harden(Object.create(oldRemotableProto)));
+
+  // Assign the arrow function to a variable to set its .name.
+  const toString = () => `[${allegedName}]`;
+  const remotableProto = harden(
+    Object.create(oldRemotableProto, {
+      toString: {
+        value: toString,
+      },
+      [Symbol.toStringTag]: {
+        value: allegedName,
+      },
+    }),
+  );
+
+  // Take a static copy of the properties.
+  const propEntries = Object.entries(props);
+  const mutateHardenAndCheck = target => {
+    // Add the snapshotted properties.
+    propEntries.forEach(([prop, value]) => (target[prop] = value));
+
+    // Set the prototype for debuggability.
+    Object.setPrototypeOf(target, remotableProto);
+    harden(remotableProto);
+
+    harden(target);
+    assertCanBeRemotable(target);
+    return target;
+  };
+
+  // Fail fast: check a fresh remotable to see if our rules fit.
+  mutateHardenAndCheck(Object.create(oldRemotableProto));
```

Comment:
Why are you adding the props own properties to `oldRemoteProto` as well? This looks like a bug to me.

Reply:
Adding to a fresh object whose prototype is `oldRemoteProto`, and checking it for failures to pass `assertCanBeRemotable`.